File: java/arrays/handofstraights/HandOfStraights.java
package arrays.handofstraights;

import java.util.TreeMap;

/**
 * Hand of straights problem. Given an array of cards, return true if the given cards can be grouped in the given 
 * group size. Also, the cards in any group needs to be consecutive.
 */

public class HandOfStraights {
    
    public boolean solve(int data[], int groupSize){

        if(data == null || data.length == 0 || groupSize < 1) return false;
        TreeMap<Integer, Integer> countsMap = new TreeMap<>();
        //Go through each element and add the element to count map with it's count.
        for (int i = 0; i < data.length; i++) {
            
            if(countsMap.containsKey(data[i])){
                countsMap.replace(data[i], countsMap.get(data[i]) + 1);
            }else{
                countsMap.put(data[i], 1);
            }
            
        }

        while(countsMap.size() > 0){
            int key = countsMap.firstKey();
            for (int i = key; i < key + groupSize; i++) {
                
                if(!countsMap.containsKey(i)) return false;
                
                if(countsMap.get(i) == 1){
                    countsMap.remove(i);
                }else{
                    countsMap.replace(i, countsMap.get(i) - 1);
                }
            }
        }

        return true;
    }

    public static void main(String[] args) {
        HandOfStraights handOfStraights = new HandOfStraights();
        System.out.println("Hand of straights problem.");
        int[] data = new int[]{1,2,3,6,2,3,4,7,8};
        int groupSize = 3;
        boolean result = handOfStraights.solve(data, groupSize);
        System.out.println("can arrange cards in group of 3 ? " + result);
    }
}

File: kotlin/arrays/handofstraights/HandOfStraights.kt
fun buildCount(treeMap: java.util.TreeMap<Int, Int>, hand: IntArray): Unit{
    for (element in hand) {
        if(treeMap.containsKey(element)){
            treeMap.put(element, treeMap.get(element)!!.plus(1))
        }else{
            treeMap.put(element, 1)
        }
    }
}

fun handOfStraights(hand: IntArray, W: Int): Boolean{
    if(hand == null || hand.size == 0 || hand.size % W != 0) return false
    
    val treeMap = java.util.TreeMap<Int, Int>()
    buildCount(treeMap, hand)

    while(treeMap.size > 0){
        val firstKey = treeMap.firstKey()
        val group = IntArray(W) { _ -> 0 }
        var index = 0
        
        for (key in firstKey until (firstKey + W)) {
            val value = treeMap.get(key)

            if(value == null) return false
            
            if(value == 1){
                group[index++] = key
                treeMap.remove(key)
            }else if(value > 1){
                group[index++] = key
                treeMap.put(key, treeMap.get(key)!!.minus(1))
            }else{
                return false
            }
        }
    }
    return true
}

fun main(args: Array<String>) {
    val hand = intArrayOf(1,2,3,6,2,3,4,7,8)
    val W = 3
    val result = handOfStraights(hand, W)
    println("Can rearrange? : $result")

    val hand1 = intArrayOf(1,2,3,4,5)
    val W1 = 4
    val result1 = handOfStraights(hand1, W1)
    println("Can rearrange? : $result1")
}